raulin0/backup-juggler
# feat: implements the command line interface

## Description
This pull request introduces significant enhancements to the Backup Juggler project, including updates to the contribution guide, MkDocs configuration, documentation content, installation instructions, project dependencies, CLI functionality, and additional tests. The changes aim to improve the overall usability, clarity, and completeness of the project.

## Changes Made
### Contribution Guide Updates:
- Added a high-level overview of the project structure.
- Detailed the purpose and functionality of each directory.
- Provided a comprehensive breakdown of the backup_juggler directory.
- Emphasized the use of the Typer and Rich libraries.
- Described the testing setup and documentation structure.
- Clarified the usage of the mkdocstrings plugin.
- Added a "Tools" section for environment management and task automation.
- Outlined the contribution process step-by-step.
- Expanded on the Git Flow workflow.
- Advised the use of Commitizen for structured commit messages.
- Added instructions for pushing changes and opening pull requests.
- Encouraged continuous improvement of the contribution guide.
### MkDocs Configuration Updates:
- Added markdown extensions for attr_list, admonition, pymdownx.tasklist, and pymdownx.superfences.
- Included an extra CSS file for customization.
- Added a custom run command in the extra section.
### Documentation Updates:
- Added comprehensive documentation for the Backup Juggler command-line tool.
- Updated the index with Backup Juggler quick start information.
- Added an initial version of the tutorial.
- Included installation instructions and styling updates.
### Build Updates:
- Updated project dependencies.
- Added a new script entry to the configuration.
### CLI Functionality:
- Added the functionality of the Backup Juggler CLI.
- Implemented the do_backups and get_size commands.
- Added a --version option to display the version.
- Implemented the --help option to show the help message.
- Styled the output using the Rich library.
### CLI Tests:
- Added additional tests for the Backup Juggler CLI.
- Covered various CLI functionalities and scenarios.
- Utilized fixtures for test setup and execution.
### Fixture:
- Added the run_backup_juggler fixture for CLI testing.

## Motivation and Context
These changes address existing issues with incomplete documentation, unclear contribution guidelines, and missing functionality in the Backup Juggler project. The updates aim to provide contributors with clear instructions, enhance the documentation structure and content, improve the CLI's functionality and test coverage, and ensure the project's dependencies are up to date.

## How to Test

1. Review the updated contribution guide and ensure the instructions are clear and complete.
2. Verify that the MkDocs configuration is updated and the documentation is generated correctly.
3. Explore the documentation to confirm that it covers all aspects of the Backup Juggler tool.
4. Follow the installation instructions and ensure the Backup Juggler CLI is installed successfully.
5. Test the CLI functionality by performing backups and calculating source sizes using different options and parameters.
6. Run the existing and new tests to ensure they cover various scenarios and produce the expected results.
7. Utilize the run_backup_juggler fixture for testing the Backup Juggler CLI with different configurations.
8. Validate the overall functionality, usability, and accuracy of the project.

## Related Issue(s)
N/A

## Checklist
- [x] Documentation has been added or updated to reflect the new functionality.

Please review and merge this pull request at your earliest convenience. Any feedback or suggestions for improvement are welcome. Thank you!
